### DISCUSSION TITLE
AAD i autoryzacja między SPA i backendem

### DISCUSSION THREAD

**Discussion by szympulka** [AzureCommunityPL/group#43] [Pomoc] [answered]
Tworzę 2 aplikacje, gdzie autoryzacja będzie opierać się o AAD. Pierwsze jest napisana w reacie a druga w net core 3.1. Aplikacje są zarejestrowane i działają poprawnie osobno. Problem pojawił mi się, gdy chciałem zrobić api cal to backendu. Skopiowałem token z frontu i wkleiłem go do postmana by zrobić strzał do bakcendu. Apka zwraca mi błąd: 

`System.UnauthorizedAccessException: IDW10201: Neither scope or roles claim was found in the bearer token. 
   at Microsoft.Identity.Web.WebApiAuthenticationBuilderExtensions.<>c__DisplayClass3_1.<<AddMicrosoftWebApi>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler1.AuthenticateAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationService.AuthenticateAsync(HttpContext context, String scheme)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)`

Jak mogę w przypadku aplikacji reacta pozyskać token z rolami?

Dockumentacja której używałem
[.net](https://www.josephguadagno.net/2020/06/12/protecting-an-asp-net-core-api-with-microsoft-identity-platform)
[react](https://docs.microsoft.com/en-us/graph/tutorials/react)

**Comment by szympulka** [comment]
Po tym jak AAD mi się rozpropagował to zaczęło prawie śmigać. Teraz dostaję już błąd 403 przy odpowiedzi serwera. Czy mogę zezwolić by token działał dla 2 różnych endpointów ?

**Comment by piotrgwiazda** [comment]
"Poprawnie" to powinno być tak, że jedna aplikacja reprezentuje API i ma
zarejestowane Api ID. Np. api://moje-api-dev. Aplikacje uzyskują dostęp do
api przez zakładkr Expose API. Wtedy każda aplikacja ma swój ID token
wydany dla użytkownika (pole audience wskazuje na ID aplikacji) natomiast
dla backendu każda aplikacja uzyskuje Access token wydany dla API (pole
audience wskazuje API ID).

Jeśli backend weryfikuje już jakoś tokeny i nie ma jak tego zmienić, to
możesz uprawnić jedną aplikację do drugiej za pomocą Expose API.

Nie jestem tylko pewnien czy aplikacja typu native może mieć zdefiniowane
API ID.

pon., 31 sie 2020, 13:36 użytkownik Szymon Pulka <notifications@github.com>
napisał:

**Comment by mgrabarz** [comment, marked as answer]
@piotrgwiazda Pełna zgoda z tym co napisałeś. Każda aplikacja "logiczna" powinna mieć swoje "App Registration" w AAD. 

Nie widzę przeciwwskazań, by aplikacje pisane w React miały swoje app registration. Należy jednak mieć na uwadze dwie rzeczy:
- React oznacza aplikację działającą w przeglądarce, czyli mamy tu do czynienia z "Public Client" w nazewnictwie OAuth. Te nie mają prawa używać secretów, bo nie ma technicznej możliwości ich ukryć
- W związku z powyższym możemy pozyskać tokeny za pomocą Implicit lub Authorization Code przy czym ten drugi musi być z rozszerzaniem PKCE. Jeśli apka chce strzelać do backendu (a taki jest opis problemu), to zgodnie z aktualnymi rekomendacjami wyklucza to użycie Implicit Flow. 

Kiedyś na blogu to opisywałem w dwóch artykułach - były front/back, b2c, tokeny, scopy, inny .NET Core ale ogólna koncepcja się nie zmieniła. 
[Link](https://grabarz.pl/tags/#openidconnect)

**Comment by piotrgwiazda** [comment]
Szczęśliwie MSAL 2.0 ze wsparciem dla PKCE jest już GA! Nadal mi się nie
podoba, że trzyma tokeny w session storage (bo cross site scripting) ale
jest lepiej.

śr., 2 wrz 2020, 09:56 użytkownik Marek Grabarz <notifications@github.com>
napisał:
